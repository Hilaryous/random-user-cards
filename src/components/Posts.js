// @flow
import React, { PureComponent } from 'react'
import styled from 'react-emotion'
import map from 'ramda/src/map'
import Card from './Card'
import Post from './Post'
import Text from './Text'
import type { PostType } from '../types/Post'
import { getPosts } from '../utils'

const PostsContainer = styled(Card)({
  gridColumn: '1 / span 3',
}, ({ modifier }) => {
  switch (modifier) {
    case 'withPosts':
      return {
        paddingTop: 40,
        paddingBottom: 80,
      }
    default:
      return {}
  }
})

type Props = {
  userId: string,
}

type State = {
  error: ?{
    message: string,
  },
  isLoaded: boolean,
  posts: Array<PostType>
}

class Posts extends PureComponent<Props, State> {
  state = {
    error: null,
    isLoaded: false,
    posts: [],
  };

  componentDidUpdate(prevProps: Props) {
    const { userId } = this.props
    const { userId: prevUserId } = prevProps
    if (userId !== prevUserId) {
      fetch(`https://jsonplaceholder.typicode.com/posts?userId=${userId}`)
        .then(res => res.json())
        .then(
          (result) => {
            this.setState(() => ({
              isLoaded: true,
              posts: getPosts(result, 5),
            }))
          },
          (error) => {
            this.setState(() => ({
              isLoaded: true,
              error,
            }))
          },
        )
    }
  }

  render() {
    const { error, isLoaded, posts } = this.state
    const { userId } = this.props
    const hasPosts = posts.length
    if (error) {
      return (
        <div>
          Error:
          {error.message}
        </div>
      )
    } if (userId && !isLoaded) {
      return (
        <div>
          Loading...
        </div>
      )
    }
    return (
      <PostsContainer modifier={hasPosts ? 'withPosts' : null}>
        { hasPosts ? (
          map(({ id, title, body }) => (
            <Post
              title={title}
              body={body}
              key={id}
            />
          ), posts)
        ) : (
          <Text modifier="center">
            Click on a card to view five of thier posts
          </Text>
        )}
      </PostsContainer>
    )
  }
}

export default Posts
