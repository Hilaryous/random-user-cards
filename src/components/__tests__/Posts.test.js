/* eslint-disable function-paren-newline */
import React from 'react'
import renderer from 'react-test-renderer'
import Posts from '../Posts'

describe('Posts', () => {
  it('shows the correct snapshot tree for the Posts component, default', () => {
    const tree = renderer.create(
      <Posts currentUserId="" />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('shows the correct snapshot tree for the Posts component, default', () => {
    const tree = renderer.create(
      <Posts currentUserId="1" />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
