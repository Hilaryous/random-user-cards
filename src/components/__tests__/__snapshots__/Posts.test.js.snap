// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts shows the correct snapshot tree for the Posts component, default 1`] = `
<li
  className="css-175b6in"
>
  <p
    className="css-19nqlgd"
  >
    Click on a card to view five of thier posts
  </p>
</li>
`;

exports[`Posts shows the correct snapshot tree for the Posts component, default 2`] = `
<li
  className="css-175b6in"
>
  <p
    className="css-19nqlgd"
  >
    Click on a card to view five of thier posts
  </p>
</li>
`;
